include/nil/blueprint/components/algebra/fields/plonk/logic_or_flag.hpp

```diff
@@ -0,0 +1,233 @@
+                    const plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>& component, 
+                    assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &assignment, 
+                    const typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::input_type &instance_input, 
+                    const std::uint32_t start_row_index){
+
+                        std::size_t row = start_row_index;
+                        std::size_t witness_amount = WitnessesAmount;
+
+                        std::array<typename BlueprintFieldType::value_type, 7> t;
+
+                        t[0] = var_value(assignment, instance_input.x);
+                        t[1] = var_value(assignment, instance_input.y);
+                        t[2] = t[0].is_zero() ? t[0] : t[0].inversed();
+                        t[3] = t[1].is_zero() ? t[1] : t[1].inversed();
+                        t[4] = t[0]*t[2];
+                        t[5] = t[1]*t[3];
+                        t[6] = t[4] + t[5] - t[4]*t[5];
+                        
+                        std::size_t _idx;
+                        for(std::size_t i=0;i<component.rows_amount;i++){
+                            for(std::size_t j=0; j < witness_amount; j++){
+                                _idx = i*witness_amount + j;
+                                assignment.witness(component.W(j), row+i) = t[_idx % 7];  // circularly fill all witness/rows table
+                            }
+                        }
+                        assignment.witness(component.W(witness_amount-1), row+component.rows_amount-1) = t[6]; //always set last element as output
+
+
+                        return typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::result_type(component, start_row_index);
+                }
+
+                template<typename BlueprintFieldType, typename ArithmetizationParams, std::uint32_t WitnessesAmount>
+                void generate_gates(
+                    const plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>& component,
+                    circuit<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &bp,
+                    assignment<crypto3::zk::snark::plonk_constraint_system<BlueprintFieldType, ArithmetizationParams>> &assignment, 
+                    const typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::input_type &instance_input, 
+                    const std::uint32_t first_selector_index
+                ){
+                    
+                    using var = typename plonk_logic_or_flag_component<BlueprintFieldType, ArithmetizationParams, WitnessesAmount>::var;
+
+                    std::size_t offset = component.rows_amount == 3 ? -1 : 0;
```

Comment:
This doesn't work for WitnessesAmount = 2: there are going to be 4 rows, but the offset is still going to start from zero, and are going to spillover into +2.

Reply:
When WitnessesAmount = 2, I redefine positions of _fx, _fy and _f and assign them a new gate with selector_index+1 and start_row_index+2 in line 172-179.